opentelemetry-sdk/src/opentelemetry/sdk/trace/sampling.py

```diff
@@ -161,26 +192,70 @@ def should_sample(
         self,
         parent_context: Optional["SpanContext"],
         trace_id: int,
-        span_id: int,
         name: str,
-        attributes: Optional[Attributes] = None,  # TODO
+        attributes: Attributes = None,  # TODO
         links: Sequence["Link"] = (),
-    ) -> "Decision":
+    ) -> "SamplingResult":
+        decision = Decision.NOT_RECORD
+        if trace_id & self.TRACE_ID_LIMIT < self.bound:
+            decision = Decision.RECORD_AND_SAMPLED
+        if decision is Decision.NOT_RECORD:
+            return SamplingResult(decision)
+        return SamplingResult(decision, attributes)
+
+    def get_description(self) -> str:
+        return "ProbabilitySampler{{{}}}".format(self._rate)
+
+
+class ParentOrElse(Sampler):
+    """
+    If a parent is set, follows the same sampling decision as the parent.
+    Otherwise, uses the delegate provided at initialization to make a
+    decision.
+
+    Args:
+        delegate: The delegate sampler to use if parent is not set.
+    """
+
+    def __init__(self, delegate: Sampler):
+        self._delegate = delegate
+
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        name: str,
+        attributes: Attributes = None,  # TODO
+        links: Sequence["Link"] = (),
+    ) -> "SamplingResult":
         if parent_context is not None:
-            return Decision(parent_context.trace_flags.sampled)
+            if (
+                not parent_context.is_valid
+                or not parent_context.trace_flags.sampled
+            ):
+                return SamplingResult(Decision.NOT_RECORD)
+            return SamplingResult(Decision.RECORD_AND_SAMPLED, attributes)
+
+        return self._delegate.should_sample(
+            parent_context=parent_context,
+            trace_id=trace_id,
+            name=name,
+            attributes=attributes,
+            links=links,
+        )
 
-        return Decision(trace_id & self.TRACE_ID_LIMIT < self.bound)
+    def get_description(self):
+        return "ParentOrElse{{{}}}".format(self._delegate.get_description())
 
 
-ALWAYS_OFF = StaticSampler(Decision(False))
+ALWAYS_OFF = StaticSampler(Decision.NOT_RECORD)
 """Sampler that never samples spans, regardless of the parent span's sampling decision."""
 
-ALWAYS_ON = StaticSampler(Decision(True))
+ALWAYS_ON = StaticSampler(Decision.RECORD_AND_SAMPLED)
 """Sampler that always samples spans, regardless of the parent span's sampling decision."""
 
-
-DEFAULT_OFF = ProbabilitySampler(0.0)
+DEFAULT_OFF = ParentOrElse(ALWAYS_OFF)
 """Sampler that respects its parent span's sampling decision, but otherwise never samples."""
 
-DEFAULT_ON = ProbabilitySampler(1.0)
+DEFAULT_ON = ParentOrElse(ALWAYS_ON)
```

Comment:
The spec doesn't mention `StaticSampler` just the two possibilities of `AlwaysOnSampler` and `AlwaysOffSampler`. It also says they MUST have `get_description()` of `Always{On,Off}Sampler`. Idk how important keeping singletons is

Reply:
The spec requires `AlwaysOn` and `AlwaysOff` to be included (so StaticSampler isn't part of the api). I think we can implement it however we want. I believe calling `get_description` on the singletons produces the desired result?

Reply:
My mistake, hadn't seen the `StaticSample` code